test/index.js

```diff
@@ -2225,6 +2416,103 @@ describe('def', function() {
+                   '‘sillyConst’ requires ‘a’ to satisfy the Alternative type-class constraint; the value at position 1 does not.\n'));
+  });
+
+  it('supports unary type variables', function() {
+    var env = $.env.concat([Maybe]);
+    var def = $.create({checkTypes: true, env: env});
+
+    //  f :: Type
+    var f = $.UnaryTypeVariable('f');
+
+    //  Functor :: TypeClass
+    var Functor = $.TypeClass(
+      'my-package/Functor',
+      function(x) { return x != null && typeof x.map === 'function'; }
+    );
+
+    //  map :: Functor f => (a -> b) -> f a -> f b
+    var map =
+    def('map',
+        {f: [Functor]},
+        [$.Function([a, b]), f(a), f(b)],
+        function(f, functor) { return functor.map(f); });
+
+    eq(map(R.inc, Nothing), Nothing);
+    eq(map(R.inc, Just(42)), Just(43));
+
+    throws(function() { map(R.inc, [1, 2, 3]); },
+           errorEq(TypeError,
+                   '‘map’ applied ‘a -> b’ to the wrong number of arguments\n' +
+                   '\n' +
+                   'map :: Functor f => (a -> b) -> f a -> f b\n' +
+                   '                     ^\n' +
+                   '                     1\n' +
+                   '\n' +
+                   'Expected one argument but received three arguments:\n' +
+                   '\n' +
+                   '  - 1\n' +
+                   '  - 0\n' +
+                   '  - [1, 2, 3]\n'));
+
+    //  Foldable :: TypeClass
+    var Foldable = $.TypeClass('my-package/Foldable', hasMethods(['reduce']));
+
+    //  sum :: Foldable f => f FiniteNumber -> FiniteNumber
+    var sum =
+    def('sum',
+        {f: [Foldable]},
+        [f($.FiniteNumber), $.FiniteNumber],
+        function(foldable) {
+          return foldable.reduce(function(x, y) { return x + y; }, 0);
+        });
+
+    eq(sum([1, 2, 3, 4, 5]), 15);
+
+    throws(function() { sum({x: 1, y: 2, z: 3}); },
+           errorEq(TypeError,
+                   'Type-class constraint violation\n' +
+                   '\n' +
+                   'sum :: Foldable f => f FiniteNumber -> FiniteNumber\n' +
+                   '       ^^^^^^^^^^    ^\n' +
+                   '                     1\n' +
+                   '\n' +
+                   '1)  {"x": 1, "y": 2, "z": 3} :: Object, StrMap Number\n' +
+                   '\n' +
+                   '‘sum’ requires ‘f’ to satisfy the Foldable type-class constraint; the value at position 1 does not.\n'));
+
+    throws(function() { sum(['foo', 'bar', 'baz']); },
+           errorEq(TypeError,
+                   'Invalid value\n' +
+                   '\n' +
+                   'sum :: Foldable f => f FiniteNumber -> FiniteNumber\n' +
+                   '                       ^^^^^^^^^^^^\n' +
+                   '                            1\n' +
+                   '\n' +
+                   '1)  "foo" :: String\n' +
+                   '\n' +
+                   'The value at position 1 is not a member of ‘FiniteNumber’.\n'));
```

Comment:
As a result of sanctuary-js/sanctuary#234 I realized we lacked tests for the two cases above. After writing the tests it was apparent that we also lacked logic to handle these two cases. I've made some changes so these tests now pass.

/cc @jdudek